### PR TITLE
Connectionmanager synchronisation refactor

### PR DESCRIPTION
**Pull request by paddybyers**
This change introduces an event queue for serialisation of all
CM-related event handling. In general, state changes, associated
notifications, and other event processing, are performed without
the CM locked, to avoid the deadlocks that have been prevalent in
earlier releases. However, transport state changes that require
a synchronous change in CM behaviour - that is, transport
disconnections - are handled synchronously so that the transport
is not used when it is no longer viable.

Fixes: https://github.com/ably/ably-java/issues/524
Fixes: https://github.com/ably/ably-java/issues/495

### CODE REVIEW

File: lib/src/main/java/io/ably/lib/transport/ConnectionManager.java [hunk 1]
```diff
@@ -1004,41 +1305,42 @@ private boolean connectImpl(StateIndication request) {
...
-			}
-			transport = null;
+		/* if connected, send an explicit close message and await response */
+		boolean isConnected = currentState.state == ConnectionState.connected;
+		if(isConnected) {
+			try {
+				Log.v(TAG, "Requesting connection close");
+				transport.send(new ProtocolMessage(ProtocolMessage.Action.close));
+				return false;
+			} catch (AblyException e) {}
```
Comment (QuintinWillison) [inline lib/src/main/java/io/ably/lib/transport/ConnectionManager.java]: This looks a bit suspect. Could you perhaps add a comment here to explain why swallowing exceptions at this point is normal behaviour?

**Comment by QuintinWillison** [comment]
I gave the realtime suite another run (it took 23m 1s) and this time I got two failures:

```
connect_invalid_recover_fail[binary_protocol]
java.lang.AssertionError: Verify correct error code is given expected:<80018> but was:<40100>
	at org.junit.Assert.fail(Assert.java:88)
	at org.junit.Assert.failNotEquals(Assert.java:834)
	at org.junit.Assert.assertEquals(Assert.java:645)
	at io.ably.lib.test.realtime.RealtimeConnectFailTest.connect_invalid_recover_fail(RealtimeConnectFailTest.java:318)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at org.junit.internal.runners.statements.FailOnTimeout$CallableStatement.call(FailOnTimeout.java:298)
	at org.junit.internal.runners.statements.FailOnTimeout$CallableStatement.call(FailOnTimeout.java:292)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.lang.Thread.run(Thread.java:748)
```

and

```
recover_disconnected[binary_protocol]
java.lang.AssertionError: Verify success from all message callbacks
	at org.junit.Assert.fail(Assert.java:88)
	at org.junit.Assert.assertTrue(Assert.java:41)
	at io.ably.lib.test.realtime.RealtimeRecoverTest.recover_disconnected(RealtimeRecoverTest.java:103)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at org.junit.internal.runners.statements.FailOnTimeout$CallableStatement.call(FailOnTimeout.java:298)
	at org.junit.internal.runners.statements.FailOnTimeout$CallableStatement.call(FailOnTimeout.java:292)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.lang.Thread.run(Thread.java:748)
```

File: lib/src/main/java/io/ably/lib/types/ConnectionDetails.java [hunk 2]
```diff
@@ -46,6 +47,9 @@ ConnectionDetails readMsgpack(MessageUnpacker unpacker) throws IOException {
 				case "maxInboundRate":
 					maxInboundRate = unpacker.unpackLong();
 					break;
+				case "maxOutboundRate":
```
[the remark this answers is not in the document]
Comment (paddybyers) [reply lib/src/main/java/io/ably/lib/types/ConnectionDetails.java]: It's unrelated, but this missing field was causing warnings when running the tests.

File: lib/src/main/java/io/ably/lib/types/ProtocolMessage.java [hunk 3]
```diff
@@ -206,6 +206,10 @@ ProtocolMessage readMsgpack(MessageUnpacker unpacker) throws IOException {
 				case "auth":
 					auth = AuthDetails.fromMsgpack(unpacker);
 					break;
+				case "connectionKey":
```
[the remark this answers is not in the document]
Comment (paddybyers) [reply lib/src/main/java/io/ably/lib/types/ProtocolMessage.java]: Same as the `maxOutboundRate` issue above

File: lib/src/test/java/io/ably/lib/test/realtime/RealtimeConnectFailTest.java [hunk 4]
```diff
@@ -54,7 +54,8 @@ public void connect_fail_notfound_error() throws AblyException {
 
 			ErrorInfo fail = connectionWaiter.waitFor(ConnectionState.failed);
 			assertEquals("Verify failed state is reached", ConnectionState.failed, ably.connection.state);
-			assertEquals("Verify correct error code is given", 404, fail.statusCode);
+			/* FIXME: sometimes this fails with 401 */
```
[the remark this answers is not in the document]
Comment (paddybyers) [reply lib/src/test/java/io/ably/lib/test/realtime/RealtimeConnectFailTest.java]: https://github.com/ably/ably-java/commit/766303473f1dc6b4ccb90bc5025373831be50509

https://github.com/ably/ably-java/issues/540

File: lib/src/test/java/io/ably/lib/test/realtime/RealtimeReauthTest.java [hunk 5]
```diff
@@ -287,10 +287,11 @@ public void reauth_fail() {
 			try {
 				Auth.TokenDetails reauthTokenDetails = ablyRealtime.auth.authorize(null, authOptions);
 				assertFalse("Expecting exception", true);
+				System.out.println("Authorize failed to throw an exception");
```
[the remark this answers is not in the document]
Comment (paddybyers) [reply lib/src/test/java/io/ably/lib/test/realtime/RealtimeReauthTest.java]: https://github.com/ably/ably-java/commit/58f53d0526bcd3ca36a0792b28fa8d1c7ed5acfa

File: lib/src/main/java/io/ably/lib/transport/ConnectionManager.java (hunk 1 above)
Comment (paddybyers) [reply lib/src/main/java/io/ably/lib/transport/ConnectionManager.java]: https://github.com/ably/ably-java/commit/fdd87ce58cdaff61d0772768086b01d4c35aae95

**Pull request merged**